astrodata/fits.py

```diff
@@ -326,6 +330,95 @@ def normalize_indices(slc, nitems):
 
     return indices, multiple
 
+# Could parametrize some naming conventions in the following two functions if
+# done elsewhere for hard-coded names like 'SCI' in future, but they only have
+# to be self-consistent with one another anyway.
+
+def wcs_to_asdftablehdu(wcs, extver=None):
+    """
+    Serialize a gWCS object as a FITS TableHDU (ASCII) extension.
+
+    The ASCII table is actually a mini ASDF file. The constituent AstroPy
+    models must have associated ASDF "tags" that specify how to serialize them.
+    """
+
+    import jsonschema
+
+    # Create a small ASDF file in memory containing the WCS object
+    # representation because there's no public API for generating only the
+    # relevant YAML subsection and an ASDF file handles the "tags" properly.
+    try:
+        af = asdf.AsdfFile({"wcs" : wcs})
+    except jsonschema.exceptions.ValidationError:
+        # (The original traceback also gets printed here)
+        raise TypeError("Cannot serialize model(s) for 'WCS' extension {}"\
+                        .format(extver or ''))
+
+    # ASDF can only dump a YAML to a binary file object, so do that and read
+    # the text back from it for storage in a FITS extension:
+    with BytesIO() as fd:
+        with af:
+            # Generate the YAML, dumping any binary arrays as text:
+            af.write_to(fd, all_array_storage='inline')
+        fd.seek(0)
+        wcstext = fd.read().decode()
```

Comment:
Some of this code should probably be spun off into non-FITS-specific routines to support serializing gWCS in other formats, but for now it works.